From 17597d01a1b67be38d60e0c8e4da1e166383a9d5 Mon Sep 17 00:00:00 2001
From: Kexy Biscuit <kexybiscuit@aosc.io>
Date: Thu, 15 Aug 2024 04:20:24 +0800
Subject: [PATCH 11/16] Bug 1906777 - [riscv64]Keep unsigned 32-bit parameter
 sign-extended in register. r=nika DONTBUILD

On riscv64, 32-bit value in 64-bit register should be sign-extended, even
when it's type is unsigned.

https://phabricator.services.mozilla.com/D215989

Co-authored-by: Lu Yahan (yahan) <yahan@iscas.ac.cn>
---
 xpcom/reflect/xptcall/md/unix/xptcinvoke_riscv64.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/xpcom/reflect/xptcall/md/unix/xptcinvoke_riscv64.cpp b/xpcom/reflect/xptcall/md/unix/xptcinvoke_riscv64.cpp
index ddd6692156..0f8b679b90 100644
--- a/xpcom/reflect/xptcall/md/unix/xptcinvoke_riscv64.cpp
+++ b/xpcom/reflect/xptcall/md/unix/xptcinvoke_riscv64.cpp
@@ -48,7 +48,9 @@ extern "C" void invoke_copy_to_stack(uint64_t* gpregs, double* fpregs,
           value = s->val.u16;
           break;
         case nsXPTType::T_U32:
-          value = s->val.u32;
+          // On RISC-V, 32-bit values need to be sign-extended in 64-bit
+          // registers, so use the signed value here.
+          value = s->val.i32;
           break;
         case nsXPTType::T_U64:
           value = s->val.u64;
-- 
2.46.0.windows.1

